hw/top_earlgrey/dv/env/sw_symbol_util_dpi.c

```diff
@@ -0,0 +1,63 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "svdpi.h"
+
+/**
+ * Extracts the addess and size of a const symbol from a SW test.
+ *
+ * Used by a testbench to modify the given symbol in an executable (elf)
+ * generated for an embedded CPU within the DUT, before it starts executing
+ * code. This function only returns the extracted address and size of the symbol
+ * using the readelf utility. Readelf comes with binutils, a package typically
+ * available on user / corp machines. If not available, the assumption is, it
+ * can be relatively easily installed. The actual job of writing the new value
+ * into the symbol is handled by the testbench invoking this function. Written
+ * in DPI-C for ease of external command invocation and string manipulation.
+ * @param elf_file Path to elf file.
+ * @param symbol A symbol in the elf file.
+ * @param addr Address of the symbol returned (by ref).
+ * @param size Size of the symbol in bytes returned (by ref).
+ * @return The result of the operation.
+ */
+extern int sw_symbol_get_addr_size(const char *elf_file, const char *symbol,
+                                   uint64_t *addr, uint64_t *size) {
+  const char *cmd_fmt =
+      "/bin/readelf -s %s | grep %s | awk \'{print $2\" \"$3}\'";
+  size_t cmd_len = strlen(cmd_fmt) + strlen(elf_file) + strlen(symbol) -
+                   4;  // Exclude '%s's.
+  char *cmd = (char *)malloc(cmd_len);
+
+  // Construct the command.
+  if (sprintf(cmd, cmd_fmt, elf_file, symbol) != cmd_len) {
+    printf("ERROR: Failed to construct cmd: \"%s\"\n", cmd);
+    return 1;
+  }
+
+  // Run the command and extract the addr and size of the symbol.
+  FILE *fp = popen(cmd, "r");
+  if (fp == NULL) {
+    printf("ERROR: Failed to run cmd: \"%s\"\n", cmd);
+    return 1;
+  }
```

Comment:
This isn't great. Rather than running readelf piped through grep piped through awk(!!) in an external process, I think it would be *much* better to use `libelf` and do this programmatically. Apart from anything else, it's much easier to programmatically figure out what happened if something goes wrong (missing ELF? symbol happens to contain a regex character? or a space?).

See verilator_memutil.cc for an example of how to load things with libelf.

Reply:
VCS (invoked on our corp machines) compiles this source. Our IT policy restricts the installation of dev packages on corp machines.

Reply:
I have reached out to them to see what can be done here (tools are run on containers, so who knows, installation requests like these might even go through), but I suppose that will take a while to get resolved. 

However, this is bound to be a problem for other corp environments as well (WD, Synopsys, etc). Considering that, it is perhaps better to keep it this way? Alternately, I can move this entire logic to a SystemVerilog package if that is more palatable.

Reply:
I agree with @rswarbrick that we really should get rid of this code. Calling a shell pipeline in C code is really fragile.

@sriyerg can you file a follow-up issue and add a TODO mentioning that until you have feedback from the internal IT? Note that libelf-dev is an OpenTitan system requirement, and has been for quite a while, i.e. we require it to be present on all machines working on OpenTitan code.

Reply:
... and I now realize that this code is not there any more in the current version of the PR, sorry for missing that. The general sentiment about fragile shell pipelines remains, but not calling it from C also removes a major concern I had. Let's continue to look for better solutions, but not block this PR on it.

Reply:
Thanks for the response. I'm not sure that moving the call to `/bin/sh` from C to SystemVerilog changes my point (that shell pipelines are icky!), but I understand what you mean about the corporate environment. Let's hope we can get that fixed soon at the Google end, and maybe we can tidy up this code then.

Reply:
I understand the concern about the shell pipes. Unfortunately, this is not just at Google, its other companies as well, which makes it a much harder problem to solve.

Reply:
Added #3838 to track this.